### PR TITLE
Let training differentiate the gated-delta recurrence

### PR DESCRIPTION
**Pull request by jhancock1975**
## Proposed changes

`gatedDeltaUpdate` runs the recurrence through a fused Metal kernel whenever
one is available. A `CustomKernel` has no VJP, so the first backward pass of a
fine-tune aborts the process:

```
MLX/ErrorHandler.swift:345: Fatal error:
[Primitive::vjp] Not implemented for CustomKernel.
```

Qwen3.5 and Qwen3-Next use gated-delta layers for three of every four layers,
so this is every LoRA fine-tune of those models rather than an edge case. It
is a fatal error rather than a thrown one, so nothing downstream can catch it
and fall back.

Two commits, separable:

1. **Let training differentiate the gated-delta recurrence.** Adds `useKernel`
   to `gatedDeltaUpdate`, defaulting to `true` so inference is untouched, and
   has the three models that call it pass `!training`. This is what the Python
   models already do: `mlx_lm/models/qwen3_5.py` passes
   `use_kernel=not self.training` for the same reason. The ops path is already
   differentiable and is already the fallback for a key dimension that is not
   a multiple of 32.

2. **Recompute gated-delta chunks on the backward pass.** The ops path keeps
   every step's state for the backward pass. At Qwen3.5-27B's shapes, 48 value
   heads of 128x128 in fp32, that is about 4 MB a step, so a thousand-token
   sequence holds gigabytes per layer. The recurrence now runs in chunks of 16
   steps through a `CustomFunction` whose backward recomputes the chunk, which
   is what `mx.checkpoint` does for the Python trainer. Take commit 1 without
   commit 2 if you would rather; commit 1 stands on its own.

### Tests

Two added to `GatedDeltaTests`:

- `testGatedDeltaOpsPathIsDifferentiable` — a backward pass through the ops
  path returns finite, non-zero gradients for q, k and v at T = 40, which
  spans more than one recompute chunk.
- `testGatedDeltaOpsPathMatchesTheKernel` — the two paths agree on the output
  and on the carried state. Compared relative to the values involved: these
  inputs are random and the recurrent state compounds to about 1.2e6 by
  T = 40, where one part in a million is 1.25 in absolute terms.

The three existing `GatedDeltaTests` still pass. Run on an M-series Mac with
the metallib colocated with the test host, since `swift test` alone cannot
load it.

### How the crash was confirmed

A backward pass through `gatedDeltaUpdate` on unpatched `main` (e3d4a20)
exits with signal 5 and the fatal error quoted above. That is why the new test
can only demonstrate the fix by passing: there is no pre-fix state in which it
fails cleanly.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx-swift-lm/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

## AI usage

- [ x] I have read this PR description in full and approve it as my own, and it
      accurately describes the code changes.
- AI usage disclosure: I used an AI assistant to help write the code in this pull request. I have created this pull request manually.

### CODE REVIEW

**Comment by davidkoski** [comment]
Very interesting -- I was thinking about this from the Cuda angle for #588 but this the same issue.

File: Libraries/MLXLMCommon/GatedDelta.swift [hunk 1]
```diff
@@ -230,6 +230,59 @@ private func gatedDeltaStepOps(
...
+                mask: mask.map { $0[0..., t] }
+            )
+            ys.append(y)
+            state = newState
+        }
+        return [MLX.stacked(ys, axis: 1), state]
+    }
+
+    // The closure holds a locked state object; two callers serialise on it.
+    nonisolated(unsafe) static let unmasked: ([MLXArray]) -> [MLXArray] = CustomFunction {
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/GatedDelta.swift:271]: This should probably produce `Sendable` function on the mlx-swift side.

Comment (davidkoski) [reply Libraries/MLXLMCommon/GatedDelta.swift:271]: https://github.com/ml-explore/mlx-swift/issues/475

**Review by davidkoski — APPROVED** [review]
Change looks good.  i filed an item for mlx-swift improvements.  Thank you!
